working/augmentation-libraries/parts_with_imports.md

```diff
@@ -0,0 +1,367 @@
+        directives from a single Dart file, to refer to that computation, which
+        introduces a scope containing the declarations introduced by all the
+        `import` s (the declarations of the export scope of each imported
+        library, minus those hidden by a `show` or `hide` operator, combined
+        such that a name conflicts of different declarations is not an error,
+        but the name is marked as conflicted in the scope, and then referencing
+        it is an error.)
+*   Let *P* be a *prefix scope* containing all the import prefixes declared by
+    the current file. The parent scope of *P* is *I*.
+    *   The *P* scope contains an entry for each name where the current file
+        has an `import` directive with that name as prefix, `as name`. (If an
+        import is `deferred`, it’s a compile-time error if more than one
+        `import` directive in the same file has that prefix name, as usual.
+        _It’s not an error if two import deferred prefixes have the same name
+        if they occur in different files, other file’s imports are only
+        suggestions._)
+    *   The *P* scope binds each such name to a *prefix import scope*,
+        *P*<sub>*name*</sub>, computed as *importsOf*(*S*<sub>*name*</sub>)
+        where *S*<sub>*name*</sub> is the set of import directives with that
+        prefix name.
+    *   If an import is `deferred`, its *P*<sub>*name*</sub> is a *deferred
+        scope* which has an extra `loadLibrary` member added, as usual, and the
+        import has an implicit `hide  loadLibrary` modifier.
+    *   If *P*<sub>*name*</sub> is not `deferred`, and the parent scope in *C*
+        has a non-deferred prefix import scope with the same name,
+        *C*<sub>*name*</sub>, then the parent scope of *P*<sub>*name*</sub> is
+        *C*<sub>*name*</sub>. _A part file can use the same prefix as a prefix
+        that it inherits, because inherited imports are only suggestions. If it
+        adds to that import scope, by importing into it, that can shadow
+        existing declarations, just like in the top-level declaration scope. A
+        deferred prefix import scope cannot be extended, and cannot extend
+        another prefix scope, deferred prefix scopes are always linked to a
+        single import directive._
+    *   _It’s possible to look further up in the import chain *C* for a prefix
+        scope to extend. Here it’s chosen that that importing parent file gets
+        to decide which names the part file has access to. If it wants to make
+        a transitive parent import prefix available, it should just not shadow
+        it._
+
+That is: The combined import scope of a Dart file is a chain of the combined
+import scopes of the file and its parent files, each step adding two scopes:
+The (unnamed, top-level) import scope of the unprefixed imports and the prefix
+scope with prefixed imports, each shadowing names further up in the chain.
```

Comment:
I am not sure about this "shadowing names further up the chain" part, presumably this is what you meant earlier when talking about being able to ignore/shadow parent imports.

Basically, your current import scope would implicitly shadow your inherited import scope? So a Foo which is directly imported, would not conflict with a Foo coming from parent imports, it would shadow it?

I don't personally like that, I would rather it just be a conflict, and have only one import scope.

Reply:
Yes. If a part introduces a name itself, by an import or import prefix, then it can ignore whether a parent had imported the same name.

If a part file only uses names it has imported itself, then it can completely ignore the inherited imports.

The inherited imports cannot cause a *name conflict* in the part file, any name introduced by the part file will always take precedence.

Reply:
Having only one import scope is where we are today.

Do you mean only one import scope per part file, which *collapses* the inherited scope and the new imports into one, making any name conflict an conflicting name?

Or do you mean any import in any part file being visible in the entire library, and any name conflict being a conflicted import in every file of the library?

The latter seems rather harsh. Not even import prefixes would be local, they too would likely be joined with import prefixes of other libraries, and name conflicts inside would be conflited imports.

The former still makes it much harder for a part or parent file, so any file, to introduce new imports. It would only conflict with imports of its parents, not any import anywhere in the library, but a part file that wants to import something, still has to do it with a prefix (which itself may already be defined) if it wants to be certain to not get a conflict.
(Because you can't hide inherited members.)

I rather think that allowing a part file to *choose* which names it defines itself, gives a good flexibility. You *can* inherit everything, and most days you want to, but you can also import your own declarations.

I can see that having a third party import with no prefix and no `show` can risk the third party library introducing a new name that shadows the name that you used to inherit from the parent file. That would become a compile-time error if we didn't allow conflicts, but without that, it could risk running and doing something wrong. (Most likely something won't type any more, but it's harder to track down the cause.)

I can see that as an argument for not allowing shadowing. (Shadowing by something that can add names at any time is dangerous. But we've been living with that for a long time, not sure this is any different.)

Reply:
Thinking about this more, it's definitely possible, but I feel like it's pointing in the opposite direction of the "inherit declartion scope of parents, so you can shadow it" idea.

For that one, instead of every library having all declarations (one global scope), declarations from parents and siblings are only inherited (from the parent) so that the part file can have imports that shadow declarations from outside of its own subtree.
This gives the part file *more* power to override inherited names.

Here, as I read it, we're arguing for making it *harder* to shadow imports inherited from the parent, something that is currently specified as always possible.
That gives the part file *less* power to override inherited names.

I can't make a good argument for why shadowing actual declarations from the current library, something the library author has complete control over, is considerd a good thing, 
but shadowing imports of other libraries, something that is notorious for containing unnecessary names, and being able to introduce new names, if not imported with `show` or `prefix`, is a bad thing.

"Shadow everything!" is  consistent and powerful.
"Shadow nothing!" is consistent, but probably too restrictive.
"Shadow imports!" (what's currently specified) considers imports less important than actual library declarations, which is how I see them. Also because you can *reimport* anything you need by adding imports to the part file.
"Shadow declarations, but not imports" ... I just can't find a good argument for that combination. 

The only benefit I see for the last combiation is that an accidental addition to your own imports not shadowing the original you thought you used, but becoming a conflcit instead.
But it allows an accidental addition to your imports to shadow a declaration of the *current library*, which is at least as bad.

I'll file an issue to discuss these combinations some more.